provider/pdns.go

```diff
@@ -0,0 +1,295 @@
+	domainFilter DomainFilter
+	// filter hosted zones by type (e.g. private or public)
+	zoneTypeFilter ZoneTypeFilter
+
+	// Swagger API Client
+	client *pgo.APIClient
+
+	// Auth context to be passed to client requests, contains API keys etc.
+	authCtx context.Context
+}
+
+// Function for debug printing
+func stringifyHTTPResponseBody(r *http.Response) (body string) {
+
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(r.Body)
+	body = buf.String()
+	return body
+
+}
+
+// NewPDNSProvider initializes a new PowerDNS based Provider.
+func NewPDNSProvider(server string, apikey string, domainFilter DomainFilter, dryRun bool) (*PDNSProvider, error) {
+
+	// Do some input validation
+
+	// We do not support dry running, exit safely instead of surprising the user
+	// TODO: Add Dry Run support
+	if dryRun {
+		log.Fatalf("PDNS Provider does not currently support dry-run, stopping.")
+	}
+
+	if server == "localhost" {
+		log.Warnf("PDNS Server is set to localhost, this is likely not what you want. Specify using --pdns-server=")
+	}
+
+	if apikey == "" {
+		log.Warnf("API Key for PDNS is empty. Specify using --pdns-api-key=")
+	}
+	if len(domainFilter.filters) == 0 {
+		log.Warnf("Domain Filter is not supported by PDNS. It will be ignored.")
+	}
+
+	provider := &PDNSProvider{}
+
+	cfg := pgo.NewConfiguration()
+	cfg.Host = server
+	cfg.BasePath = server + apiBase
+
+	// Initialize a single client that we can use for all requests
+	provider.client = pgo.NewAPIClient(cfg)
+
+	// Configure PDNS API Key, which is sent via X-API-Key header to pdns server
+	provider.authCtx = context.WithValue(context.TODO(), pgo.ContextAPIKey, pgo.APIKey{Key: apikey})
+
+	return provider, nil
+}
+
+func convertRRSetToEndpoints(rr pgo.RrSet) (endpoints []*endpoint.Endpoint, _ error) {
+	endpoints = []*endpoint.Endpoint{}
+
+	for _, record := range rr.Records {
+		//func NewEndpointWithTTL(dnsName, target, recordType string, ttl TTL) *Endpoint
+		endpoints = append(endpoints, endpoint.NewEndpointWithTTL(rr.Name, record.Content, rr.Type_, endpoint.TTL(rr.Ttl)))
+	}
+
+	return endpoints, nil
+}
+
+// Zones returns the list of all zones controlled by the pdns server as a list of pdns Zone structs
+func (p *PDNSProvider) Zones() (zones []pgo.Zone, _ error) {
+	zones, _, err := p.client.ZonesApi.ListZones(p.authCtx, defaultServerID)
+	if err != nil {
+		log.Warnf("Unable to fetch zones. %v", err)
+		return nil, err
+	}
+
+	return zones, nil
+
+}
+
+// convertEndpointsToZones marshals endpoints into pdns compatible Zone structs
+func (p *PDNSProvider) convertEndpointsToZones(endpoints []*endpoint.Endpoint, changetype pdnsChangeType) (zonelist []pgo.Zone, _ error) {
+	/* eg of mastermap
+	    { "example.com":
+		{ "app.example.com":
+		    { "A": ["192.168.0.1", "8.8.8.8"] }
+		    { "TXT": ["\"heritage=external-dns,external-dns/owner=example\""] }
+		}
+	    }
+	*/
+	mastermap := make(map[string]map[string]map[string][]*endpoint.Endpoint)
```

Comment:
is it possible to refactor this code, so that we don't have triple nested maps ?

Reply:
@ffledgling could you take a look on that again?

Reply:
I have a local copy of this patched, but this is blocked on mock testing which is exhibiting segfaults, I'm trying to still fix this. :+1: